### PR TITLE
MM-54502 - Update regex to force first character to be alpha

### PR DESCRIPTION
**Pull request by sbishel**
#### Summary

This PR updates the server side regex to enforce the first character of a username to be alphabetic. Matching the client-side validation. 

The only real change occurs in `server/public/model/user.go`. The remaining changes are to make various tests conform to that validation. 

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-54502

#### Release Note
```release-note
User APIs now enforce username beginning with alphabetic character, matching client-side validation.
```

### CODE REVIEW

**Comment by sbishel** [comment]
After seeing all the different tests that needed fixed for this, I worry this could be considered a breaking change to the API. Could this break other client apps? plugins? or does it matter?

File: server/channels/api4/apitestlib.go [hunk 1]
```diff
@@ -459,15 +459,21 @@ func (th *TestHelper) InitLogin() *TestHelper {
 }
 
 func (th *TestHelper) InitBasic() *TestHelper {
+	mlog.Debug("INITBASIC")
```
Comment (isacikgoz) [inline server/channels/api4/apitestlib.go:462]: Are these logs forgotten?

**Review by isacikgoz — CHANGES_REQUESTED** [review]
There are some logs needs to be removed, apart from that why we are preventing usernames to start with numeric characters? Also, what would happen to usernames before this change?

**Comment by isacikgoz** [comment]
Thanks for the info, I think we should only be concerned existing usernames. Tools like https://github.com/mattermost/mattermost-load-test-ng are relying on server side verification. Also we may never know what current users have as usernames. If they created users/bots from CLI, it will still fall under the server side verification. So, eg. importing users may have some exclusions on this regard. Thoughts?

**Comment by hmhealey** [comment]
> @hmhealey was this causing an issue somewhere or just something you noticed?

It's mostly something I noticed. I know we've made some assumptions about usernames starting with letters in the code, but it does seem like most places handle it fine. The only place I've 100% found an issue is that at-mentions on mobile won't render correctly for those users because the pattern makes that assumption (https://github.com/mattermost/commonmark.js/blob/master/lib/inlines.js#L949)

**Comment by sbishel** [comment]
@hmhealey @isacikgoz @streamer45 - I'm thinking this is a case where the cure is worse than the disease.  This fix is essentially making a breaking change to our APIs. As Ibrahim mentioned above, load test tools rely on server side verification, so they would probably fail and need updated. I would guess this will also cause issues in the wild. 

Thoughts on closing this as "won't fix" or making it with next major version?

**Comment by hmhealey** [comment]
Yeah, that makes sense. I agree that it's a technically a breaking change, so we should leave it until the next major version at least. While there's that one issue I found caused by this, we can always fix that independently or just treat it as a known issue and fix it if anyone actually runs into it in the wild.

**Comment by isacikgoz** [comment]
Agreed on postponing to major release, thanks @sbishel!

**Comment by hmhealey** [comment]
Related report: https://github.com/mattermost/mattermost-mobile/issues/4850

**Comment by hmhealey** [comment]
Unassigning myself until we're ready to merge this. Feel free to reassign when it's ready for merging or if we want to review it early

**Comment by sbishel** [comment]
@hmhealey @isacikgoz @streamer45 - The window for v10.0 has opened. Is this breaking change we still want to take otherwise I will close.

**Comment by hmhealey** [comment]
Yes, I think it would be good to include this for 10.0 since it'll help ensure that the state of the app stays correct. While it has slight potential to be a breaking change for someone, I think it's more likely that we'd run into a weird bug report for a case that seemingly shouldn't happen if we leave were to leave this unfixed

**Review by hmhealey — APPROVED** [review]

**Review by streamer45 — APPROVED** [review]

**Comment by streamer45** [comment]
@sbishel Not sure if this deserves a quick security pass, given that, technically speaking, we are dealing with credentials.

**Comment by sbishel** [comment]
> @sbishel Not sure if this deserves a quick security pass, given that, technically speaking, we are dealing with credentials.

Added security, but not sure they will have time.

**Review by isacikgoz — APPROVED** [review]
lgtm, thanks for your patience @sbishel 💪

**Review by esarafianou — APPROVED** [review]

**Comment by NadavTasher** [comment]
This PR breaks users synced with LDAP - if their username begins with a number (i.e 1234username) a new username is randomized on every LDAP sync, making it impossible to mention these users in posts.

**Pull request merged**